### ISSUE TITLE
[DevQ] Is Autofac and Automapper needed?

### ISSUE THREAD

**Issue by chibidev** [Jackett/Jackett#1508]
This is a development inquiry rather than an issue or any other kind of question. The reason I'm asking whether the two nuggets are really necessary is because when I initially started contributing to Jackett the hardest part was to get used to Autofac and Automapper. In my perspective neither of those solve any code maintenance issues, however they do create some and make it really hard to follow code, therefore I see no value in them.
However I understand if those are necessary for Jackett to work (although I would really like to understand why; there might be some cases that I missed, I just don't understand their value right now). In that case as both of those are outdated (Autofac is more recent, however it was updated approx. 6 months ago), I'd suggest updating them and modify the code to use the newer, non-static API of Autofac.

Also, is there any other place where I can pose such questions (e.g. Gitter)? Or is the conversation here on GitHub fine?

**Comment by chibidev** [comment]
Okay, Autofac is getting on my nerves... It randomly started reporting exceptions (did not happen before, seriously don't understand why), so I needed to **manually exclude** the types I created from the registered types... This is another example that poses me the question: is it really necessary?

**Comment by kaso17** [comment]
I can't tell you why it was used in the first time but I agree that they causes more problems than they solve.
Last time I checked neither of them (autofac/automapper) were compatible with .NET core.
I would suggest that we welcome any patches to reduce/get get rid of autofac/automapper usage.

@flightlevel what's your opinion on this?

Regarding communication:
There's a slack chat but it's not being used. I personally would suggest that we continue such discussions as github issues (better indexing/archival than slack/gitter).

**Comment by flightlevel** [comment]
Yep, agree with kaso, this is the best place for discussion as we are a community project and everyone should be able to see/throw their opinion in. And it makes it easier having everything in the one place

With Autofac, I can't give you a great answer as I came along after it was all going. Is it specifically Autofac that you don't like or more just too much dependency injection?

.NET Core 2 should be less than a month away. My suggestion would be to wait for that and then we can then see how much work it will be to port. If not too bad, this would be the ideal time IMO to make some major breaking changes like above and culling some code eg. remove Curlsharp, replace owin with Kestrel, look at adding Selenium so we would have a headless browser etc

**Comment by chibidev** [comment]
Thank you both for the answers.

The problem with all the dependency frameworks I came along is that they usually just hide issues and cause others, rather than really solving anything.
In my experience they are used in either of the following:
1. There are too many dependencies for an object, and the constructor requires a plethora of input params. However this is usually caused by wrong design and so I would rather refactor the class hierarchy rather than just hide the construction with a different object that magically knows everything. It usually indicates that an object controls/owns/does a lot of things that it probably shouldn't. And if there is true reason behind having 5+ input params, I think they should be there and should be explicitly stated 
2. There are "proxy" objects that intialize other objects (e.g. Factories) and they require all the members/possible implementations dependencies. This once again is probably caused by wrong design. The proxy object is hiding the reason why these dependencies are required, and that makes it really hard to understand why the long constructor member list. 
3. Dependency-injected objects that may initialize other dependency-injected objects need to have all the members' dependencies. This is BS in itself. If you're doing dependency injection, do it everywhere, or don't do it. If you mix it up, you might face such problems.
4. There are things to be passed to everything (e.g. a logger) and don't want to manually include that everywhere. This is probably the most sane reason, however there's a good chance you can do that in another way. If there's really something that everything uses, using a global object is not satanic. However if you don't want to, there's no problem providing that to everything: it is a true dependency.

There are several problems they create. Usually there will be a single, static lifetime object that handles magically everything, which is alarming in itself (at least for me). They handle things in ways you don't know, and it really makes it hard to know the lifetime of your objects. I come from a hard C++ background, so I might be too hung up on knowing the lifetime of the objects, however in my experience it really makes it easy to understand and reason about code, which is probably the most important aspect of maintainability for me (and determines whether someone needs 1 week, 5 months, or a year to fully understand a product).

I must confess I don't adore dependency injection altogether (my problem is basically that you have to modify the code/API/architecture in order to make it testable, and this hurts my code artist conscious), but I see the value behind it. In languages like C# that has tradition passing interfaces around in the constructor rather than allocating the object inside, DI is acceptable (although having a language feature for that would be much simpler and more reasonable, but not a lot of languages support that), DI is part of the language/standard library/tradition, and it is OK to follow, and rendering the code easier to test automatically produces great value. I just don't like when frameworks/libraries hide that, because of false reasons (like above), and I miss out a lot of information when I read the code (1st rule of ~~clean code~~ communication: state everything explicitly, so it won't be misunderstood).

**Comment by chibidev** [comment]
@flightlevel I understand the reason, and if you guys want to wait for core 2.0 I have no problem supporting it. However I want to say that in my experience doing these things in iterations is probably more reasonable, instead of doing a single-shot global refactoring. Both doing and reviewing it is easier step-by-step. An if we start doing those replacements, we'll probably finish them by the time core 2.0 is out, and it will probably make the transition easier.
Just my 2 cents.

**Comment by cadatoiva** [comment]
Automapper is almost completely removed, one more class pair to go and we can dump it.

Autofac is a different animal right now. I think a large part would be shifting some IService classes into a static class implementation with threading support. That one is going to take some time and thinking to get done.

**Comment by garfield69** [comment]
closing as stale.
